### DISCUSSION TITLE
moving release.config.js to .config folder not reading configuration correctly. version 23.0.0

### DISCUSSION THREAD

**Discussion by coderatchet** [semantic-release/semantic-release#3158] [Configuration]
### Current behavior

When running a release, the config is not being picked up even though the breaking changes section says explicitly that i should move the `release.config.js` file into the `.config` folder.

with a file `.config/release.config.js` containing a branch configuration:

```javascript
module.exports = {
  branches: ['main'],
  // ...
}
```

When running `semantic-release`, it has an error saying it must have 1-3 release branches configured:

```
A minimum of 1 and a maximum of 3 release branches are required in the branches configuration (https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#branches).

This may occur if your repository does not have a release branch, such as master.

Your configuration for the problematic branches is [].
```

#### To Reproduce:

Have a basic node library with a minimum `release.config.js` file in the .config folder and branches configured to ['main']

run the `semantic-release --dry-run` (with local repo and branches if necessary)

This works fine if the `release.config.js` file is in the project's root directory, which is also unexpected behaviour

### Expected behavior

Expect for the release to pick up the correct `.config/release.config.js` file as described in the [breaking-changes of version 23.0.0](https://github.com/semantic-release/semantic-release/releases/tag/v23.0.0)

### `semantic-release` version

23.0.0

### CI environment

buildkite

### Plugins used

commit-analyzer, release-notes-generator, npm, github, git

### `semantic-release` configuration

```javascript
module.exports = {
  branches: ['main'],
  dryRun: false,
  repositoryUrl: 'https://github.com/XXX',
  plugins: [
    [
      '@semantic-release/commit-analyzer',
      {
        releaseRules: [{ type: 'chore', scope: 'deps', release: 'patch' }],
      },
    ],
    [
      '@semantic-release/release-notes-generator',
      {
        preset: 'conventionalcommits',
        presetConfig: {
          types: [
            { type: 'feat', section: 'Features' },
            { type: 'fix', section: 'Bug Fixes' },
            { type: 'build', section: 'Dependencies' },
            { type: 'chore', hidden: true },
            { type: 'docs', hidden: true },
            { type: 'style', hidden: true },
            { type: 'refactor', hidden: true },
            { type: 'perf', section: 'Performance Improvements' },
            { type: 'test', hidden: true },
            { type: 'ci', hidden: true },
          ],
        },
      },
    ],
    '@semantic-release/npm',
    '@semantic-release/github',
    [
      '@semantic-release/changelog',
      {
        changelogFile: 'CHANGELOG.md',
      },
    ],
    [
      '@semantic-release/git',
      {
        assets: ['package.json', 'yarn.lock', 'CHANGELOG.md'],
        message: 'chore(release): ${nextRelease.version} [skip ci]\n\n${nextRelease.notes}',
      },
    ],
  ],
};
```

### CI logs

ERELEASEBRANCHES The release branches are invalid in the `branches` configuration.
--
  | [2024-01-23T06:04:30Z] [ci:release] A minimum of 1 and a maximum of 3 release branches are required in the branches configuration (https://github.com/semantic-release/semantic-release/blob/master/docs/usage/configuration.md#branches).
  | [2024-01-23T06:04:30Z] [ci:release]
  | [2024-01-23T06:04:30Z] [ci:release] This may occur if your repository does not have a release branch, such as master.
  | [2024-01-23T06:04:30Z] [ci:release]
  | [2024-01-23T06:04:30Z] [ci:release] Your configuration for the problematic branches is [].
  | [2024-01-23T06:04:30Z] [ci:release]
  | [2024-01-23T06:04:30Z] [ci:release] AggregateError:
  | [2024-01-23T06:04:30Z] [ci:release]     SemanticReleaseError: The release branches are invalid in the `branches` configuration.
  | [2024-01-23T06:04:30Z] [ci:release]         at default (file:///opt/app/node_modules/semantic-release/lib/get-error.js:6:10)
  | [2024-01-23T06:04:30Z] [ci:release]         at file:///opt/app/node_modules/semantic-release/lib/branches/index.js:45:19
  | [2024-01-23T06:04:30Z] [ci:release]         at Array.reduce (<anonymous>)

**Comment by travi** [comment]
could you switch to using a `.releaserc.js` instead? if so, does that result in different behavior? this breaking change is the result of upgrading cosmiconfig, as mentioned in our release notes. could you raise an issue there to get a better understanding of what might be causing your issue?

**Reply by coderatchet** [reply]
Unfortunately, Renaming the file to `.config/releaserc.js` did not work. Is this still an indicator that it's to do with cosmiconfig?

**Reply by travi** [reply]
sorry, there are enough variables to this that i should have been more clear. 

i've never used a `release.config.js` with semantic-release, so i admit that i'm not entirely clear on where the disconnect is. the way i read the migration instructions from cosmiconfig, i thought moving an existing `release.config.js` from the root of the repo into the `.config/` directory would be enough to adjust to the change.

i have used a `.releaserc.js` in some cases where i need the config to be dynamic and need the `.js` format. that file should be able to stay in the root of the repo instead of going in the `.config/` directory (but note the `.` at the beginning of the file. unless you have a specific reason for the `*.config.js` filename format, i think the `.releaserc.js` alternative should accomplish the same outcome. 

could you try moving the alternative name back to the repo root and add the `.` to the beginning of the file. interested to at least confirm that it would be an option to get back to working.

**Reply by mirkolenz** [reply]
I just had the same issue when upgrading to semantic release v23: I moves the config to `.config/release.config.js` and got the error about not having defined any branches. I stumbled upon this discussion and changed the file name to `.releaserc.js` and everting seems to be working just fine. Maybe it would be a good idea to adjust the manual and remove the instructions about `release.config.js` until a solution has been found? Thanks for maintaining this excellent piece of software and being responsive about questions like these ❤️

**Reply by travi** [reply]
Thank you for sharing the confirmation. Happy to review a PR with that adjustment.

**Reply by mirkolenz** [reply]
Great, I will give it a shot! Should I just remove mentions of the `release.config.js` file altogether?

Another question while I am at it: Does semantic-release support the files `.releaserc.mjs` and `.releaserc.ts` as well? I just skimmed over the cosmicconfig repo and saw that they mention support for these files. If this is possible, I would add that information to the docs as well.

**Reply by travi** [reply]
we use cosmicconfig directly, so any formats that it enables should be available to users of semantic-release. given that, maybe it is better to link to the cosmiconfig docs for that info rather than trying to replicate in the semantic-release docs?

**Comment by amenella** [comment]
Hi, I'm facing the same issue since upgrade to v23.

My setup is a GitlabCI pipeline (in a monorepo) which is launching the semantic-release cli using some args (specific for each subrepo - defined by the CI pipeline) **and** a dynamic JS config for semantic-release (which contains common parameters such as plugins etc...)

Historically, the config was named `release.config.js` and located at the root of the monorepo, since upgrade to v23, I tried with `release.config.cjs`, `.releaserc.js`, `.config/.releaserc.js` and none of them is working (therefore semantic-release is considering its default config including npm/github plugins).

Everything is working fine with v22, so I'm sticking with this version for the moment.

**Reply by travi** [reply]
the context of the monorepo sounds like it is introducing additional complexity here. since we dont officially support using semantic-release with monorepos, i'm afraid we cant justify the effort to investigate this complexity ourselves. if community members are able to discover a path forward, we welcome the sharing of that knowledge. i would recommend working with the cosmiconfig maintainers to better understand how the breaking changes may have resulted in the behavior you are seeing in your project

**Reply by amenella** [reply]
hi, thanks @travi for your answer, the context of the monorepo seems not related with the current issue (the semantic-release config is located at the root of the repo like any "normal" setup)

but maybe you were refering to the fact that we are using both a config and cli params at the same time (thus semantic-release is favoring cli params over config)?

anyway, it is working for us with v22, we'll stay on this one for the moment

**Reply by gojanpaolo** [reply]
We're also in a monorepo and only v22 worked.
Opened a [ticket on semantic-release-monorepo](https://github.com/pmowrer/semantic-release-monorepo/issues/153).

**Reply by gojanpaolo** [reply]
EDIT: Workaround below didn't actually work
---
This seemed to work on our end with v23.
In root package.json
```
{
  ...
  "scripts": {
    ...
    "semantic-release": "yarn lerna exec --stream \"yarn semantic-release -e $(pwd)/release.config.js\""
  }
}
```
The relevant part is passing `-e $(pwd)/release.config.js` to `semantic-release`.

**Reply by gojanpaolo** [reply]
The workaround with `-e` flag didn't actually work because v23 is not calling `semantic-release-monorepo` plugin. We downgraded to v22 for now.

e.g. for `analyzeCommits` step, v23 does this

```
[semantic-release] › ℹ  Start step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
[semantic-release] [@semantic-release/commit-analyzer] › ℹ  Analyzing commit: ...
```

And with v22
```
[semantic-release] › ℹ  Start step "analyzeCommits" of plugin "[Function: semantic-release-monorepo]"
[semantic-release] [[Function: semantic-release-monorepo]] › ℹ  Start step "analyzeCommits" of plugin "@semantic-release/commit-analyzer"
```

**Reply by peikk0** [reply]
This is also discussed in https://github.com/semantic-release/semantic-release/discussions/3163

**Comment by iamstarkov** [comment]
it probably should be a proper issue/bug instead of community discussion

**Comment by alper98** [comment]
Can confirm that `v22` works, but `v23` doenst. 

It doesnt recognize any of my configuration files, no matter the extension or the location of the file.

**Comment by bradennapier** [comment]
Have y'all tried `.cjs` that still works for me in root directory (havent tried moving to `.config` directory)

```
release.config.cjs
```

**Comment by cdaringe** [comment]
My issue was that my `package.json` had incorrect repository URL. Another thing for y'all to check for the same error messaging from OP.

**Reply by timnolte** [reply]
Thank you so much as this was exactly my problem. We have a separate private source repository from our public distribution repository and that was exactly the key.
